sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveMetricView.scala

```diff
@@ -0,0 +1,350 @@
+ *      WHERE product = 'product_1' GROUP BY region
+ * }}}
+ *
+ * The rule operates on unresolved plans and transforms [[MetricViewPlaceholder]] nodes
+ * into resolved logical plans that can be further optimized and executed.
+ */
+case class ResolveMetricView(session: SparkSession) extends Rule[LogicalPlan] {
+  private def parser: ParserInterface = session.sessionState.sqlParser
+  override def apply(plan: LogicalPlan): LogicalPlan = {
+    if (!plan.containsPattern(METRIC_VIEW_PLACEHOLDER)) {
+      return plan
+    }
+    plan.resolveOperatorsUp {
+      // CREATE PATH: to create a metric view, we need to analyze the metric view
+      // definition and get the output schema (with column metadata). Since the measures
+      // are aggregate functions, we need to use an Aggregate node and group by all
+      // dimensions to get the output schema.
+      case mvp: MetricViewPlaceholder if mvp.isCreate && mvp.child.resolved =>
+        val (dimensions, measures) = buildMetricViewOutput(mvp.desc)
+        Aggregate(
+          // group by all dimensions
+          dimensions.map(_.toAttribute).toSeq,
+          // select all dimensions and measures to get the final output (mostly data types)
+          (dimensions ++ measures).toSeq,
+          mvp.child
+        )
+
+      // SELECT PATH: to read a metric view, user will use the `MEASURE` aggregate function
+      // to read the measures, so it'll lead to an Aggregate node. This way, we only need to
+      // Resolve the Aggregate node based on the metric view output and then replace
+      // the AttributeReference of the metric view output to the actual expressions.
+      case node @ MetricViewReadOperation(metricView) =>
+        // step 1: parse the metric view definition
+        val (dimensions, measures) =
+          parseMetricViewColumns(metricView.outputMetrics, metricView.desc.select)
+
+        // step 2: build the Project node containing the dimensions
+        val dimensionExprs = dimensions.map(_.namedExpr)
+        // Drop the source columns if it conflicts with dimensions
+        val sourceOutput = metricView.child.output
+        // 1. hide the column conflict with dimensions
+        // 2. add an alias to the source column so they are stable with DeduplicateRelation
+        // 3. metric view output should use the same exprId
+        val dimensionAttrs = metricView.outputMetrics.filter(a =>
+          dimensions.exists(_.exprId == a.exprId)
+        )
+        val sourceProjList = sourceOutput.filterNot { attr =>
+          // conflict with dimensions
+          dimensionAttrs
+            .resolve(Seq(attr.name), session.sessionState.conf.resolver)
+            .nonEmpty
+        }.map { attr =>
+          if (attr.metadata.contains(MetricViewConstants.COLUMN_TYPE_PROPERTY_KEY)) {
+            // no alias for metric view column since the measure reference needs to use the
+            // measure column in MetricViewPlaceholder, but an alias will change the exprId
```

Comment:
`MetricViewPlaceholder.output` is fully decoupled with its child. To not change the exprId we need to add an alias to retain the original exprId, and you are doing the opposite. The `Project` should be constructed like this:
```
Project(
  Seq(
    Alias(source_attr_1, name)(exprId = output_metric_1.exprId),
    ...
  ),
  source
)
```